src/service/mod.rs

```diff
@@ -34,98 +53,187 @@ pub fn enable(time_out: i32) -> Result<ServiceState, String> {
+    match state {
+        State::Inactive => Ok(state),
+        State::Deactivating => handler(time_out, State::Inactive),
+        State::Failed => Err(Error::Failed),
         _ => {
-            let mut message = dbus_message!(SD_SERVICE_MANAGER,
-                                            SD_SERVICE_PATH,
-                                            SD_MANAGER_INTERFACE,
-                                            "StopUnit");
-            message.append_items(&["NetworkManager.service".into(), "fail".into()]);
-            dbus_connect!(message);
-
-            wait(time_out, ServiceState::Inactive)
+            let message = try!(Message::new_method_call(SD_SERVICE_MANAGER,
+                                                        SD_SERVICE_PATH,
+                                                        SD_MANAGER_INTERFACE,
+                                                        "StopUnit")
+                    .map_err(Error::Message))
+                .append2("NetworkManager.service", "fail");
+
+            let connection = try!(Connection::get_private(BusType::System)
+                .map_err(Error::Connection));
+
+            try!(connection.send_with_reply_and_block(message, 2000).map_err(Error::Connection));
+
+            handler(time_out, State::Inactive)
+        }
+    }
+}
+
+/// Gets the state of the Network Manager service.
+///
+/// # Examples
+///
+/// ```
+/// use network_manager::service;
+/// let state = service::state().unwrap();
+/// println!("{:?}", state);
+/// ```
+pub fn state() -> Result<State, Error> {
+    let message = try!(Message::new_method_call(SD_SERVICE_MANAGER,
+                                                SD_SERVICE_PATH,
+                                                SD_MANAGER_INTERFACE,
+                                                "GetUnit")
+            .map_err(Error::Message))
+        .append1("NetworkManager.service");
+
+    let connection = try!(Connection::get_private(BusType::System).map_err(Error::Connection));
+
+    let response = try!(connection.send_with_reply_and_block(message, 2000)
+        .map_err(Error::Connection));
+
+    let path = try!(response.get1::<Path>().ok_or(Error::NotFound));
+
+    let response = try!(Props::new(&connection,
+                                   SD_SERVICE_MANAGER,
+                                   path,
+                                   SD_UNIT_INTERFACE,
+                                   2000)
+        .get("ActiveState")
+        .map_err(Error::Props));
+
+    try!(response.inner::<&str>().ok().ok_or(Error::NotFound)).parse()
+}
+
+fn handler(time_out: u64, target_state: State) -> Result<State, Error> {
+    if time_out == 0 {
+        return state();
+    }
+
+    let timer =
+        Timer::default().sleep(Duration::from_secs(time_out)).then(|_| Err(Error::TimedOut));
+
+    let process = CpuPool::new_num_cpus().spawn_fn(|| {
+        let connection = try!(Connection::get_private(BusType::System).map_err(Error::Connection));
```

Comment:
the reason I didn't is that it will result in a big, compile time, string concat function which is unneccessary + those strings are only used in the `handler` function